### PR TITLE
added OdhApplication, OdhDocument K8 ResourceTypes

### PR DESCRIPTION
**Pull request by kkasravi**
- [x] Created docs/README.md describing the changes 
- [x] updated README.md, CONTRIBUTING.md to show how to deploy OdhApplication, OdhDocuent resources
- [x] updated frontend/src, backend/src to fetch/watch OdhApplication, OdhDocument k8 resources
- [x] moved data/{applications,documents} to install/deploy/base/{odh-application.yaml,odh-document.yaml} 
- [x] added CRDs for OdhApplication, OdhDocument to install/deploy/base

### CODE REVIEW

File: backend/src/routes/api/quickstarts/quickStartUtils.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { getComponentFeatureFlags } from '../../../utils/features';
 
 const quickStartsGroup = 'console.openshift.io';
 const quickStartsVersion = 'apiextensions.k8s.io/v1';
-const quickStartsPlural = 'rhodsquickstarts';
+const quickStartsPlural = 'odhquickstarts';
```
Comment (jeff-phillips-18) [inline backend/src/routes/api/quickstarts/quickStartUtils.ts:10]: Why this change?

File: backend/src/types.ts [hunk 2]
```diff
@@ -87,44 +87,59 @@ export type KubeFastifyInstance = FastifyInstance & {
  * Common types, should be kept up to date with frontend types
  */
 
-export type OdhApplication = {
-  metadata: {
-    name: string;
```
Comment (jeff-phillips-18) [inline backend/src/types.ts:92]: I believe name is required and should remain in the definition.

File: backend/src/types.ts [hunk 3]
```diff
@@ -87,44 +87,59 @@ export type KubeFastifyInstance = FastifyInstance & {
...
+export interface OdhApplication {
+  apiVersion?: string;
+  kind?: string;
+  metadata?: {
+    [k: string]: unknown;
```
Comment (jeff-phillips-18) [inline backend/src/types.ts:94]: Do we need these catch alls?

File: backend/src/utils/componentUtils.ts [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 import { OdhApplication, K8sResourceCommon, KubeFastifyInstance, RouteKind } from '../types';
 
+/*
 type RoutesResponse = {
   body: {
     items: RouteKind[];
   };
   response: IncomingMessage;
 };
+*/
```
Comment (jeff-phillips-18) [inline backend/src/utils/componentUtils.ts:12]: Let's just remove this rather than comment out.

File: backend/src/utils/resourceUtils.ts [hunk 5]
```diff
@@ -80,60 +76,76 @@ const fetchInstalledKfdefs = async (fastify: KubeFastifyInstance): Promise<KfDef
...
-        const doc: OdhDocument = jsYaml.load(
-          fs.readFileSync(path.join(normalizedPath, file), 'utf8'),
-        );
-        if (doc.spec.featureFlag) {
-          if (featureFlags[doc.spec.featureFlag]) {
-            docs.push(doc);
-          }
-          return;
-        }
-        if (!doc.spec.appName || appDefs.find((def) => def.metadata.name === doc.spec.appName)) {
```
Comment (jeff-phillips-18) [inline backend/src/utils/resourceUtils.ts:120]: This prevents adding docs for applications that are feature flagged out. I think we still need to handle this.

File: backend/tsconfig.json [hunk 6]
```diff
@@ -4,16 +4,36 @@
     "baseUrl": "./src",
     "module": "commonjs",
     "esModuleInterop": true,
-    "target": "es6",
     "noImplicitAny": true,
     "moduleResolution": "node",
-    "sourceMap": true,
```
Comment (jeff-phillips-18) [inline backend/tsconfig.json:10]: why this change?

File: backend/tsconfig.json [hunk 7]
```diff
@@ -4,16 +4,36 @@
...
+    "noImplicitReturns": false,
+    "noImplicitThis": true,
+    "noUnusedLocals": false,
+    "noUnusedParameters": false,
+    "resolveJsonModule": true,
+    "strict": true,
+    "strictNullChecks": false,
+    "strictPropertyInitialization": false,
+    "stripInternal": true,
+    "target": "ES2017"
```
Comment (jeff-phillips-18) [inline backend/tsconfig.json:36]: Why this change?

File: data/quickstarts/create-aikit-notebook-quickstart.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+          #### Verify that your AI Kit notebook launched with a Python 3 kernel:
+          Is Python 3 displaying in the upper right notification bar of your notebook?
+        failedTaskHelp:
+          This task isn’t verified yet. Try the task again.
+      summary:
+        success: You have added health checks to your sample app!
+        failed: Try the steps again.
+  conclusion: You are now able to launch AI Kit notebooks and write Python code.  If you want to learn how to deploy
+    a model, take the next quick start.
+  nextQuickStart: [deploy-python-model]
```
Comment (jeff-phillips-18) [inline data/quickstarts/create-aikit-notebook-quickstart.yaml:63]: I think we want to add AIKit as its own PR.

File: frontend/src/types.ts [hunk 9]
```diff
@@ -2,44 +2,59 @@
  * Common types, should be kept up to date with backend types
  */
 
-export type OdhApplication = {
-  metadata: {
-    name: string;
```
Comment (jeff-phillips-18) [inline frontend/src/types.ts:7]: name is required and should be defined here

File: frontend/src/types.ts [hunk 10]
```diff
@@ -2,44 +2,59 @@
...
+export interface OdhApplication {
+  apiVersion?: string;
+  kind?: string;
+  metadata?: {
+    [k: string]: unknown;
```
Comment (jeff-phillips-18) [inline frontend/src/types.ts:9]: again, I don't think we need the catch alls.

File: README.md [hunk 11]
```diff
@@ -10,12 +10,25 @@ A dashboard for Open Data Hub components.
...
+      ```
+      oc apply -f install/deploy/base/custom-resource-definition.yaml
+      ```
+
+   2. Deploy the OdhApplications, OdhDocuments Resources to your OC_PROJECT
+      ```
+      oc project $OC_PROJECT
+      cat install/deploy/base/{odh-application.yaml,odh-document.yaml} | oc apply -f -
+      ```
+
```
Comment (jeff-phillips-18) [inline README.md:31]: The Makefile should be updated to do this and the scripts updated where necessary. Currently the `deploy.sh` and `undeploy.sh` scripts will not work.

**Comment by jeff-phillips-18** [comment]
@kkasravi How/when are all the docs and apps deployed to the cluster (not in dev mode)?

File: backend/src/routes/api/quickstarts/quickStartUtils.ts (hunk 1 above)
Comment (kkasravi) [reply backend/src/routes/api/quickstarts/quickStartUtils.ts]: reverted

File: backend/src/types.ts (hunk 2 above)
Comment (kkasravi) [reply backend/src/types.ts:92]: i've updated OdhApplication, OdhDocument using the types generated from kpt [typegen](https://pkg.go.dev/github.com/googlecontainertools/kpt-functions-sdk/go/cmd/typegen) which will generate k8 types for a given cluster including the CRDs. I can add typegen into the build process at a later point.

File: backend/src/types.ts (hunk 3 above)
Comment (kkasravi) [reply backend/src/types.ts]: fixed using typegen

File: backend/tsconfig.json (hunk 6 above)
Comment (kkasravi) [reply backend/tsconfig.json:10]: reverted

File: data/quickstarts/create-aikit-notebook-quickstart.yaml (hunk 8 above)
Comment (kkasravi) [reply data/quickstarts/create-aikit-notebook-quickstart.yaml]: removed

File: frontend/src/types.ts (hunk 9 above)
Comment (kkasravi) [reply frontend/src/types.ts:7]: full schema for metadata is included using kpt typegen. See prior comment from me.

File: frontend/src/types.ts (hunk 10 above)
Comment (kkasravi) [reply frontend/src/types.ts]: fixed using kpt's typegen which will generate the full metadata schema

File: backend/tsconfig.json (hunk 7 above)
Comment (kkasravi) [reply backend/tsconfig.json]: mostly reverted.

File: backend/src/utils/resourceUtils.ts (hunk 5 above)
Comment (kkasravi) [reply backend/src/utils/resourceUtils.ts:120]: added this logic

**Comment by jeff-phillips-18** [comment]
The `downstream_changes` branch was merged upstream. This PR should now go against upstream master. Sorry for the run around 😞

**Comment by kkasravi** [comment]
Hi @jeff-phillips-18 
> The downstream_changes branch was merged upstream. This PR should now go against upstream master. Sorry for the run around 😞

Does that mean i should open a PR against opendatahub-io/odh-dashboard master?
I'll close this and open a PR against opendatahub-io/odh-dashboard master. If that's the wrong target, let me know.  
I'm close to getting the make targets fixed as you noted above. I can ping you in the new PR when it's ready to review again.
Sorry about the PR size. Thanks for all your input thus far.

**Comment by jeff-phillips-18** [comment]
> Does that mean i should open a PR against opendatahub-io/odh-dashboard master?

Yes, thanks.
